src/parser/bgp/messages.rs

```diff
@@ -45,30 +45,22 @@ pub fn parse_bgp_message(
     have the smallest value required, given the rest of the
     message.
     */
-    let length = data.get_u16();
+    let length = data.read_u16()?;
     if !(19..=4096).contains(&length) {
-        return Err(ParserError::ParseError(format!(
-            "invalid BGP message length {}",
-            length
-        )));
+        return Err(ParserError::InvalidBgpMessageLength(length));
     }
 
+    // TODO: Why do we sometimes change our length estimate?
     let bgp_msg_length = if (length as usize) > total_size {
         total_size - 19
     } else {
         length as usize - 19
     };
 
-    let msg_type: BgpMessageType = match BgpMessageType::try_from(data.get_u8()) {
-        Ok(t) => t,
-        Err(_) => {
-            return Err(ParserError::ParseError(
-                "Unknown BGP Message Type".to_string(),
-            ))
-        }
-    };
+    let msg_type: BgpMessageType = BgpMessageType::try_from(data.read_u8()?)?;
 
     if data.remaining() != bgp_msg_length {
+        // TODO: Why is this not a hard error?
```

Comment:
It should be a hard error. We should also make sure we can continue parsing other messages after this (it probably does that).

Reply:
When I change this to be an error, it fails `parser::bmp::tests::test_route_monitoring`. Do you have any ideas why this happens? All of the other tests passed, including the integration tests I added which parse through an entire update file and RIB dump to check for errors.
`ParseError(InconsistentFieldLength { name: "BGP message length", expected: 109, found: 110 })`

Reply:
Honestly, I don't have much ideas on why this happens. The test was created by dumping some BMP raw messages from RouteViews BMP stream. This might indicate some potential issues with the BMP parsing that we previously were unaware of. I wonder if you could successfully run the `real-time-routeviews-kafka-openbmp` example with your updated code? That example will pull live data from RouteViews Kafka BMP stream.